cmd/provisioner-localpv/app/config.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/pkg/errors"
+	"k8s.io/api/core/v1"
+)
+
+const (
+	KeyPVBasePath     = "BasePath"
+	KeyPVRelativePath = "RelativePath"
+	KeyPVAbsolutePath = "AbsolutePath"
+)
+
+const (
+	// BetaStorageClassAnnotation represents the beta/previous StorageClass annotation.
+	// It's currently still used and will be held for backwards compatibility
+	BetaStorageClassAnnotation = "volume.beta.kubernetes.io/storage-class"
+)
+
+//CASConfigParser creates a new CASConfigPVC struct by
+// parsing and merging the configuration provided in the PVC
+// annotation - cas.openebs.io/config with the
+// default configuration of the provisioner.
+func (p *Provisioner) CASConfigParser(pvName string, pvc *v1.PersistentVolumeClaim) (*CASConfigPVC, error) {
+
+	pvConfig := p.defaultConfig
+
+	//TODO Fetch the SC and its configuration
+	scName := GetStorageClassName(pvc)
+
+	// extract the cas volume config from pvc
+	pvcCASConfigStr := pvc.ObjectMeta.Annotations[string(mconfig.CASConfigKey)]
+	if len(strings.TrimSpace(pvcCASConfigStr)) != 0 {
+		pvcCASConfig, err := cast.UnMarshallToConfig(pvcCASConfigStr)
+		if err == nil {
+			pvConfig = cast.MergeConfig(pvcCASConfig, pvConfig)
+		}
+	}
+
+	pvConfigMap, err := cast.ConfigToMap(pvConfig)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read configuration for pvc %v", pvc.ObjectMeta.Name)
+	}
+
+	c := &CASConfigPVC{
+		pvName:  pvName,
+		pvcName: pvc.ObjectMeta.Name,
+		scName:  scName,
+		config:  pvConfigMap,
+	}
+	return c, nil
+}
+
+//GetPath returns a valid PV path based on the configuration
+// or an error. The Path is constructed using the following rules:
+// If AbsolutePath is specified return it. (Future)
+// If PVPath is specified, suffix it with BasePath and return it. (Future)
+// If neither of above are specified, suffix the PVName to BasePath
+//  and return it
+// Also before returning the path, validate that path is safe
+//  and matches the filters specified in StorageClass.
+func (c *CASConfigPVC) GetPath() (string, error) {
+	//This feature need to be supported with some more
+	// security checks are in place, so that rouge pods
+	// don't get access to node directories.
+	//absolutePath := c.getConfigValue(KeyPVAbsolutePath)
+	//if len(strings.TrimSpace(absolutePath)) != 0 {
+	//	return c.validatePath(absolutePath)
+	//}
+
+	basePath := c.getConfigValue(KeyPVBasePath)
+	if len(strings.TrimSpace(basePath)) == 0 {
+		return "", fmt.Errorf("configuration error, no base path was specified")
+	}
+
+	//This feature need to be supported after the
+	// security checks are in place.
+	//pvRelPath := c.getConfigValue(KeyPVRelativePath)
+	//if len(strings.TrimSpace(pvRelPath)) == 0 {
+	//	pvRelPath = c.pvName
+	//}
+
+	pvRelPath := c.pvName
+	path := filepath.Join(basePath, pvRelPath)
+
+	return c.validatePath(path)
+}
+
+//getConfigValue is a utility function to extract the value
+// of the `key` from the ConfigMap object - which is
+// map[string]interface{map[string][string]}
```

Comment:
Can we have additional comments on how the config will look like & how the key looks like

Reply:
added an example.